// play-audio/dist/webm/WebmConstant.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebmElements = exports.DataReturn = exports.DataType = void 0;
var DataType;
(function (DataType) {
    DataType[DataType["master"] = 0] = "master";
    DataType[DataType["string"] = 1] = "string";
    DataType[DataType["uint"] = 2] = "uint";
    DataType[DataType["binary"] = 3] = "binary";
    DataType[DataType["float"] = 4] = "float";
})(DataType = exports.DataType || (exports.DataType = {}));
exports.DataReturn = {
    string: (buf) => buf.toString('utf-8'),
    uint: (buf) => parseInt(buf.toString('hex'), 16),
    float: (buf) => parseFloat(buf.toString('utf-8')),
};
exports.WebmElements = {
    "1a45dfa3": { name: "ebml", type: DataType.master },
    "4286": { name: "ebmlVersion", type: DataType.uint, return: exports.DataReturn.uint },
    "42f7": { name: "ebmlReadVersion", type: DataType.uint, return: exports.DataReturn.uint },
    "42f2": { name: "ebmlMaxIDLength", type: DataType.uint, return: exports.DataReturn.uint },
    "42f3": { name: "ebmlMaxSizeLength", type: DataType.uint, return: exports.DataReturn.uint },
    "4282": { name: "docType", type: DataType.string, return: exports.DataReturn.string },
    "4287": { name: "docTypeVersion", type: DataType.uint, return: exports.DataReturn.uint },
    "4285": { name: "docTypeReadVersion", type: DataType.uint, return: exports.DataReturn.uint },
    "18538067": { name: "segment", type: DataType.master },
    "114d9b74": { name: "seekHead", type: DataType.master },
    "4dbb": { name: "seek", type: DataType.master },
    "53ab": { name: 'seekId', type: DataType.binary },
    "53ac": { name: 'seekPosition', type: DataType.uint, return: exports.DataReturn.uint },
    "1549a966": { name: "info", type: DataType.master },
    "4489": { name: "duration", type: DataType.float, return: exports.DataReturn.float },
    "4d80": { name: "muxingApp", type: DataType.string, return: exports.DataReturn.string },
    "5741": { name: "writingApp", type: DataType.string, return: exports.DataReturn.string },
    "1f43b675": { name: "cluster", type: DataType.master },
    "e7": { name: "clusterTimecode", type: DataType.uint, return: exports.DataReturn.uint },
    "a3": { name: "simpleBlock", type: DataType.binary },
    "1654ae6b": { name: "tracks", type: DataType.master },
    "ae": { name: "trackEntry", type: DataType.master },
    "d7": { name: "trackNumber", type: DataType.uint, return: exports.DataReturn.uint },
    "83": { name: 'trackType', type: DataType.uint, return: exports.DataReturn.uint },
    "86": { name: "codecID", type: DataType.string, return: exports.DataReturn.string },
    "e1": { name: "audio", type: DataType.master },
    "b5": { name: "samplingFrequency", type: DataType.float, return: exports.DataReturn.float },
    "9f": { name: "channels", type: DataType.uint, return: exports.DataReturn.uint },
    "6264": { name: "bitDepth", type: DataType.uint, return: exports.DataReturn.uint },
    "1c53bb6b": { name: "cues", type: DataType.master },
    "bb": { name: "cuePoint", type: DataType.master },
    "b3": { name: "cueTime", type: DataType.uint, return: exports.DataReturn.uint },
    "b7": { name: "cueTrackPositions", type: DataType.master },
    "f7": { name: "cueTrack", type: DataType.uint, return: exports.DataReturn.uint },
    "f1": { name: "cueClusterPosition", type: DataType.uint, return: exports.DataReturn.uint },
};
