// play-audio/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebmElements = exports.WebmHeader = exports.WebmDemuxer = exports.OggDemuxer = exports.OpusEncoder = exports.OpusDecoder = void 0;
var opus_1 = require("./opus");
Object.defineProperty(exports, "OpusDecoder", { enumerable: true, get: function () { return opus_1.OpusDecoder; } });
Object.defineProperty(exports, "OpusEncoder", { enumerable: true, get: function () { return opus_1.OpusEncoder; } });
var ogg_1 = require("./ogg");
Object.defineProperty(exports, "OggDemuxer", { enumerable: true, get: function () { return ogg_1.OggDemuxer; } });
var webm_1 = require("./webm");
Object.defineProperty(exports, "WebmDemuxer", { enumerable: true, get: function () { return webm_1.WebmDemuxer; } });
Object.defineProperty(exports, "WebmHeader", { enumerable: true, get: function () { return webm_1.WebmHeader; } });
Object.defineProperty(exports, "WebmElements", { enumerable: true, get: function () { return webm_1.WebmElements; } });
